source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -385,164 +357,109 @@ bool Filter::parseTlvs(const std::vector<uint8_t>& tlvs) {
-        ENVOY_LOG(debug, "failed to read proxy protocol (insufficient data)");
+    ssize_t addr_len = addr_len_opt.value();
+    uint8_t upper_byte = buf[PROXY_PROTO_V2_HEADER_LEN - 2];
+    uint8_t lower_byte = buf[PROXY_PROTO_V2_HEADER_LEN - 1];
+    ssize_t hdr_addr_len = (upper_byte << 8) + lower_byte;
+    if (hdr_addr_len < addr_len) {
+      ENVOY_LOG(debug,
+                "incorrect address length, address length = {}, the expected address length = {}",
+                hdr_addr_len, addr_len);
+      return ReadOrParseState::Error;
+    }
+    // waiting for more data if there is no enough data for address.
+    if (raw_slice.len_ >= static_cast<size_t>(PROXY_PROTO_V2_HEADER_LEN + addr_len)) {
+      // The TLV remain, they are parsed in `parseTlvs()` which is called from the
+      // parent (if needed).
+      if (parseV2Header(buf)) {
+        return ReadOrParseState::Done;
+      } else {
         return ReadOrParseState::Error;
       }
-      if (ssize_t(buf_off_) + nread >= PROXY_PROTO_V2_HEADER_LEN + addr_len) {
-        ssize_t missing = (PROXY_PROTO_V2_HEADER_LEN + addr_len) - buf_off_;
-        const auto read_result = io_handle.recv(buf_ + buf_off_, missing, 0);
-        if (!read_result.ok() || read_result.return_value_ != uint64_t(missing)) {
-          ENVOY_LOG(debug, "failed to read proxy protocol (remote closed)");
-          return ReadOrParseState::Error;
-        }
-        buf_off_ += read_result.return_value_;
-        // The TLV remain, they are read/discard in parseExtensions() which is called from the
-        // parent (if needed).
-        if (parseV2Header(buf_)) {
-          return ReadOrParseState::Done;
-        } else {
-          return ReadOrParseState::Error;
-        }
-      } else if (nread != 0) {
-        const auto result = io_handle.recv(buf_ + buf_off_, nread, 0);
-        nread = result.return_value_;
-        if (!result.ok()) {
-          ENVOY_LOG(debug, "failed to read proxy protocol (remote closed)");
-          return ReadOrParseState::Error;
-        }
-        buf_off_ += nread;
-      }
-    } else {
-      // continue searching buf_ from where we left off
-      for (; search_index_ < buf_off_ + nread; search_index_++) {
-        if (buf_[search_index_] == '\n' && buf_[search_index_ - 1] == '\r') {
-          if (search_index_ == 1) {
-            // This could be the binary protocol. It cannot be the ascii protocol
-            header_version_ = InProgress;
-          } else {
-            header_version_ = V1;
-            search_index_++;
-          }
+    }
+  } else {
+    // continue searching buffer from where we left off
+    for (; search_index_ < raw_slice.len_; search_index_++) {
+      if (buf[search_index_] == '\n' && buf[search_index_ - 1] == '\r') {
+        if (search_index_ == 1) {
+          // This could be the binary protocol. It cannot be the ascii protocol.
+          // reach here due to there is no enough data for v2 protocol signature.
```

Comment:
`There is not enough data to determine if it contains the v2 protocol signature, so wait for more data.`

Reply:
Thanks!